flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServer.java

```diff
@@ -0,0 +1,411 @@
+	 * within a job archive. This class does not share code with the {@link JobArchiveFetcherTask} since the former
+	 * fetches entire archives at a given location based on a path whereas this class only fetches one specific file
+	 * based on a REST URL.
+	 */
+	public static class JobFileFetcher {
+		private final Path webDir;
+		private final Map<String, Path> cachedArchives;
+
+		public JobFileFetcher(Path webDir, Map<String, Path> cachedArchives) {
+			this.webDir = webDir;
+			this.cachedArchives = cachedArchives;
+		}
+
+		public void fetchFileForRestRequest(String requestPath) throws IOException {
+			/**
+			 * Here we extract the job ID from the request path. We are either missing a file that resides
+			 * under /jobs/:jobid/* or /jobs/:jobid.json. In the latter case indexOf will
+			 * return -1 so we manually set 'to' to the index before '.json'.
+			 */
+			int from = 6;
+			int to = requestPath.indexOf('/', 7);
+			if (to == -1) {
+				to = requestPath.length() - 5;
+			}
+			String jobID = requestPath.substring(from, to);
+			Path sourcePath = new Path(new Path(cachedArchives.get(jobID), jobID), requestPath);
+			try {
+				FileCache.copy(sourcePath, new Path(webDir, requestPath), false);
+			} catch (Exception e) {
+				LOG.debug("Failed to retrieve file {} for job {}. This may indicate an incomplete or corrupt job archive.", sourcePath, e);
+				LOG.error("Failed to retrieve file {} for job {}. This may indicate an incomplete or corrupt job archive.", sourcePath, jobID);
+			}
+		}
+	}
+
+	// =================================================================================================================
+	// File generation
+	// =================================================================================================================
+	private static FileWriter createOrGetFile(File folder, String name) throws IOException {
+		File file = new File(folder, name + ".json");
+		if (!file.exists()) {
+			Files.createFile(file.toPath());
+		}
+		FileWriter fr = new FileWriter(file);
+		return fr;
+	}
+
+	private void createDashboardConfigFile() throws IOException {
+		try (FileWriter fw = createOrGetFile(webDir, "config")) {
+			fw.write(DashboardConfigHandler.createConfigJson(webRefreshIntervalMillis));
+			fw.flush();
+		} catch (IOException ioe) {
+			LOG.error("Failed to write config file.");
+			throw ioe;
+		}
+	}
+
+	/**
+	 * This method replicates the JSON response that would be given by the {@link CurrentJobsOverviewHandler} when
+	 * listing both running and finished jobs.
+	 * 
+	 * Every job archive contains a joboverview.json file containing the same structure. Since jobs are archived on
+	 * their own however the list of finished jobs only contains a single job.
+	 * 
+	 * For the display in the HistoryServer WebFrontend we have to combine these overviews.
+	 */
+	private static void updateJobOverview(File webDir) {
```

Comment:
Does anything speak against using an actual handler that is updated with the counts when new jobs are fetched? I think that could be simpler than re-reading the file and re-wrting it.